src/backend/cuda/kernel/reduce_by_key.hpp

```diff
@@ -0,0 +1,732 @@
+    }
+
+    //
+    // Reduce each warp by key
+    char all_eq = (k == shfl_down_sync(0xFFFFFFFF, k, 1));
+    if (all_sync(0xFFFFFFFF,
+                 all_eq)) {  // check special case of single key per warp
+        v = reduce(v, shfl_down_sync(FULL_MASK, v, 1));
+        v = reduce(v, shfl_down_sync(FULL_MASK, v, 2));
+        v = reduce(v, shfl_down_sync(FULL_MASK, v, 4));
+        v = reduce(v, shfl_down_sync(FULL_MASK, v, 8));
+        v = reduce(v, shfl_down_sync(FULL_MASK, v, 16));
+    } else {
+        // preform reduction for each of the unique keys
+        int eq_check = (unique_id == shfl_down_sync(0xFFFFFFFF, unique_id, 1));
+        int update_key =
+            eq_check && (laneid < 31) &&
+            ((tidx + 1) <
+             n);  // checks if this thread should perform a reduction
+        unsigned shflmask = ballot_sync(
+            0xFFFFFFFF,
+            update_key);  // obtains mask of all threads that should be reduced
+        shflmask |= (shflmask << 1);  // shifts mask to include source threads
+                                      // that should participate in _shfl
+        compute_t<To> uval =
+            shfl_down_sync(shflmask, v,
+                           1);  // shfls data from neighboring threads
+        compute_t<To> init = Binary<compute_t<To>, op>::init();
+        v = reduce(v, (update_key ? uval : init));  // update if thread
+                                                    // requires it
+        eq_check = (unique_id == shfl_down_sync(0xFFFFFFFF, unique_id, 2));
+        update_key =
+            eq_check && (laneid < 30) && update_key && ((tidx + 2) < n);
+        shflmask = ballot_sync(0xFFFFFFFF, update_key);
+        shflmask |= (shflmask << 2);
+        uval = shfl_down_sync(shflmask, v, 2);
+        v    = reduce(v, (update_key ? uval : init));
+
+        eq_check = (unique_id == shfl_down_sync(0xFFFFFFFF, unique_id, 4));
+        update_key =
+            eq_check && (laneid < 28) && update_key && ((tidx + 4) < n);
+        shflmask = ballot_sync(0xFFFFFFFF, update_key);
+        shflmask |= (shflmask << 4);
+        uval = shfl_down_sync(shflmask, v, 4);
+        v    = reduce(v, (update_key ? uval : init));
+
+        eq_check = (unique_id == shfl_down_sync(0xFFFFFFFF, unique_id, 8));
+        update_key =
+            eq_check && (laneid < 24) && update_key && ((tidx + 8) < n);
+        shflmask = ballot_sync(0xFFFFFFFF, update_key);
+        shflmask |= (shflmask << 8);
+        uval = shfl_down_sync(shflmask, v, 8);
+        v    = reduce(v, (update_key ? uval : init));
+
+        eq_check = (unique_id == shfl_down_sync(0xFFFFFFFF, unique_id, 16));
+        update_key =
+            eq_check && (laneid < 16) && update_key && ((tidx + 16) < n);
+        shflmask = ballot_sync(0xFFFFFFFF, update_key);
+        shflmask |= (shflmask << 16);
+        uval = shfl_down_sync(shflmask, v, 16);
+        v    = reduce(v, (update_key ? uval : init));
```

Comment:
Lines 315 to 362  can be condensed into a for loop.
```cpp
        compute_t<To> init = Binary<compute_t<To>, op>::init();
        int eq_check, update_key;
        unsigned shflmask;
#pragma unroll
        for (int delta = 1; delta < 32; delta <<= 1) {
            eq_check = (unique_id == shfl_down_sync(0xFFFFFFFF, unique_id, delta));

            // checks if this thread should perform a reduction
            update_key = eq_check && (laneid < (32-delta)) && ((tidx + delta) < n);

            // obtains mask of all threads that should be reduced
            shflmask = ballot_sync(0xFFFFFFFF, update_key);

            // shifts mask to include source threads that should participate in _shfl
            shflmask |= (shflmask << delta);

            // shfls data from neighboring threads
            compute_t<To> uval = shfl_down_sync(shflmask, v, delta);

            // update if thread requires it
            v = reduce(v, (update_key ? uval : init));
        }
```

Reply:
I like this, way more readable.